File: Twopointers/MaxNumberOfKSumPairs.java
package Twopointers;

import java.util.Arrays;

public class MaxNumberOfKSumPairs {
    public int maxOperations(int[] nums, int k) {
        Arrays.sort(nums);
        int left = 0;
        int right = nums.length - 1;
        int res = 0;

        while (left < right) {
            if (nums[left] + nums[right] == k) {
                res++;
                left++;
                right--;

            } else if (nums[left] + nums[right] < k) {
                left++;
            } else {
                right--;
            }

        }
        return res;
    }

}

File: Twopointers/MaxNumberOfKSumPairs.scala
object Solution {
  def maxOperations(nums: Array[Int], k: Int): Int = {
    var left = 0
    var right = nums.length - 1
    var count = 0
    val sorted = nums.sorted

    while (left < right) {

      if (sorted(left) + sorted(right) == k) {
        left = left + 1
        right = right - 1
        count = count + 1
      } else if (sorted(left) + sorted(right) < k) {
        left = left + 1
      } else {
        right = right - 1
      }
    }

    count
  }
}
